src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PENamedTypeSymbol.cs

```diff
@@ -1807,6 +1877,160 @@ public override TypeKind TypeKind
+            foundMarkerMethod = null;
+
+            var isByRefLike = module.HasIsByRefLikeAttribute(_handle);
+            if (!isByRefLike)
+            {
+                return false;
+            }
+
+            foreach (var field in module.GetFieldsOfTypeOrThrow(_handle))
+            {
+                if ((module.GetFieldDefFlagsOrThrow(field) & FieldAttributes.Static) == 0)
+                {
+                    return false;
+                }
+            }
+
+            try
+            {
+                foreach (var methodHandle in module.GetMethodsOfTypeOrThrow(_handle))
+                {
+                    var methodName = module.GetMethodDefNameOrThrow(methodHandle);
+                    if (methodName == WellKnownMemberNames.ExtensionMarkerMethodName)
+                    {
+                        if (foundMarkerMethod != null)
+                        {
+                            return false;
+                        }
+
+                        foundMarkerMethod = methodHandle;
+                    }
+                }
+
+                return foundMarkerMethod is not null;
+            }
+            catch (BadImageFormatException)
+            {
+            }
+
+            return false;
+        }
+
+        private void DecodeExtensionType(out TypeSymbol underlyingType, out ImmutableArray<NamedTypeSymbol> baseExtensions)
+        {
+            // PROTOTYPE we should also decode `implicit` vs. `explicit`
+            // PROTOTYPE consider optimizing/caching
+
+            TypeSymbol? foundUnderlyingType;
+            if (!tryDecodeExtensionType(out foundUnderlyingType, out baseExtensions))
+            {
+                if (foundUnderlyingType is null)
+                {
+                    foundUnderlyingType = new UnsupportedMetadataTypeSymbol();
+                }
+
+                if (baseExtensions.IsDefault)
+                {
+                    baseExtensions = ImmutableArray<NamedTypeSymbol>.Empty;
+                }
+            }
+
+            underlyingType = foundUnderlyingType;
+
+            return;
+
+            bool tryDecodeExtensionType([NotNullWhen(true)] out TypeSymbol? underlyingType, out ImmutableArray<NamedTypeSymbol> baseExtensions)
+            {
+                var found = IsExtensionType(out MethodDefinitionHandle? foundMarkerMethod);
+                Debug.Assert(found);
+                var markerMethod = foundMarkerMethod!.Value;
+                var moduleSymbol = this.ContainingPEModule;
+
+                underlyingType = null;
+                baseExtensions = default;
+
+                MethodAttributes localFlags;
+                moduleSymbol.Module.GetMethodDefPropsOrThrow(markerMethod, name: out _, implFlags: out _, out localFlags, rva: out _);
+                // PROTOTYPE do we want to tighten the flags check further?
+                if ((localFlags & MethodAttributes.Private) == 0 ||
+                    (localFlags & MethodAttributes.Static) == 0)
```

Comment:
Should we use this condition as a filter in ```IsExtensionType```?

Reply:
I don't have a strong rule for what should be in `IsExtensionType` versus what should be in `DecodeExtensionType`. The way I divided is that the checked that involve cracking the cracking the method are done in `IsExtensionType`.

The only requirement is that any checks that relate to parameter types should be delayed (so we don't risk creating cycles in `TypeKind` computation) .
So technically, all the other checks (flags, by-value, etc) could be moved earlier.

I'm leaning to leave as-is, but don't feel strongly. Let me know if you have better reasons.